### ISSUE TITLE
synchronizing on OAuth.class is problematic

### ISSUE THREAD

**Issue by eve-gatt** [burberius/eve-esi#78]
https://github.com/burberius/eve-esi/blob/0ad7e172f66e528ad7ca1409adbf5436dbfb389c/src/main/java/net/troja/eve/esi/auth/OAuth.java#L119

Hi, the synchronized block here is causing problems for me, I wondered if anyone could explain the history of it? I suspect what's happening is the auth token request is not responding at https://github.com/burberius/eve-esi/blob/0ad7e172f66e528ad7ca1409adbf5436dbfb389c/src/main/java/net/troja/eve/esi/auth/OAuth.java#L224 but it's causing everything else to queue up behind it.

Given that ACCESS_TOKEN_CACHE is a ConcurrentHashMap, could we not remove the synchronized block? 

Or perhaps set a timeout on the HttpURLConnection?

**Comment by GoldenGnu** [comment]
No, we only want to refresh/update the access token once, that is why that part is synchronized.
We have to wait until it's refreshed or we can't update, anyway.

Could you explain what kind of problems you're getting with it?

**Comment by burberius** [comment]
The synchronized block should start at 
https://github.com/burberius/eve-esi/blob/0ad7e172f66e528ad7ca1409adbf5436dbfb389c/src/main/java/net/troja/eve/esi/auth/OAuth.java#L125
and use the AccessTokenData of that connection only.

We can do that change after #75 as everything is prepared already.

@eve-gatt Do you use the library in production?

**Comment by GoldenGnu** [comment]
I'm going to look into this.
I think we need a stronger link between the refresh token and the access token, and the updating of both

**Comment by eve-gatt** [comment]
> No, we only want to refresh/update the access token once

Yeah, that makes sense, but that synchronized block means that only 1 thread can update any access token at a time doesn't it?

> Could you explain what kind of problems you're getting with it?

Yeah, I suspect the http call is never returning and so all my other threads get queued behind it. I'm trying this for now to see if it helps: 
```            con.setConnectTimeout(10000);
            con.setReadTimeout(10000);
```

>@eve-gatt Do you use the library in production?

In that I've got about 40 or so corpies using it daily, yes.

**Comment by eve-gatt** [comment]
Specifically, the error I got at 03:17 this morning was:
```
Mar 04 03:17:29 tuskers app/web.1:  WARNING: Thread Thread[worker-pilot-tracker-5,5,main] has been blocked for 60090 ms, time limit is 60000 ms 
Mar 04 03:17:29 tuskers app/web.1:  io.vertx.core.VertxException: Thread blocked 
Mar 04 03:17:29 tuskers app/web.1:  	at net.troja.eve.esi.auth.OAuth.getAccessTokenData(OAuth.java:118) 
Mar 04 03:17:29 tuskers app/web.1:  	at net.troja.eve.esi.auth.OAuth.applyToParams(OAuth.java:47) 
Mar 04 03:17:29 tuskers app/web.1:  	at net.troja.eve.esi.ApiClient.updateParamsForAuth(ApiClient.java:834) 
Mar 04 03:17:29 tuskers app/web.1:  	at net.troja.eve.esi.ApiClient.invokeAPI(ApiClient.java:711) 
Mar 04 03:17:29 tuskers app/web.1:  	at net.troja.eve.esi.api.LocationApi.getCharactersCharacterIdLocation(LocationApi.java:93) 
```

**Comment by burberius** [comment]
Default Timeout is very likely 300s, so that connection just returns very late, but too late for vertx.
I hope your timeout changes work as expected!

The change to [OkHttp](http://square.github.io/okhttp/) probably also helps in this case.

**Comment by eve-gatt** [comment]
I thought the default timeout was 0, i.e. infinity. The vertx part is just a warning and I could extend that if there was a valid case for a thread taking > 1min. Let's see how setting an explicit timeout works. Unfortunately this occurs so infrequently that it's difficult to test. I'm not even sure how to simulate it.

**Comment by burberius** [comment]
So I would just say lets wait for it and if it still exists, try another fix.
Simulating it would mean you have to drop or slow down those special packages but I think that's way too complicated.

**Comment by GoldenGnu** [comment]
Work-in-progress PR (https://github.com/burberius/eve-esi/pull/79) up for this.

**Comment by GoldenGnu** [comment]
Released in 3.0.0
